File: src/server/api/routers/member.ts
import { z } from 'zod'
import { createTRPCRouter, protectedProcedure } from '../trpc'
import { TRPCError } from '@trpc/server'
import type { Member, PaginatedResponse } from '@/types/prisma'
import { logger } from '@/utils/logger' // Assuming you have a logger utility

const API_URL = process.env.PYTHON_API_URL || 'http://localhost:8000'

export const memberRouter = createTRPCRouter({
    getAll: protectedProcedure
        .input(
            z.object({
                search: z.string().optional(),
                page: z.number().default(1),
                limit: z.number().default(20)
            })
        )
        .query(async ({ input }): Promise<PaginatedResponse<Member>> => {
            logger.info('Fetching all members', { input })
            const params = new URLSearchParams({
                page: input.page.toString(),
                limit: input.limit.toString()
            })
            if (input.search) params.append('search', input.search)

            const response = await fetch(`${API_URL}/api/members?${params}`)
            if (!response.ok) {
                logger.error('Failed to fetch members', { status: response.status })
                throw new TRPCError({
                    code: 'INTERNAL_SERVER_ERROR',
                    message: 'Failed to fetch members'
                })
            }
            const data = await response.json()
            logger.info('Successfully fetched members', { count: data.items.length })
            return data
        }),

    getById: protectedProcedure.input(z.number()).query(async ({ input }): Promise<Member> => {
        logger.info('Fetching member by ID', { id: input })
        const response = await fetch(`${API_URL}/api/members/${input}`)
        if (!response.ok) {
            logger.error('Member not found', { id: input, status: response.status })
            throw new TRPCError({
                code: 'NOT_FOUND',
                message: 'Member not found'
            })
        }
        const data = await response.json()
        logger.info('Successfully fetched member', { id: input })
        return data
    }),

    create: protectedProcedure
        .input(
            z.object({
                name: z.string(),
                email: z.string().email(),
                address: z.string().optional()
            })
        )
        .mutation(async ({ input }): Promise<Member> => {
            logger.info('Creating new member', { input })
            const response = await fetch(`${API_URL}/api/members`, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify(input)
            })
            if (!response.ok) {
                const error = await response.json()
                logger.error('Failed to create member', { input, error })
                throw new TRPCError({
                    code: 'BAD_REQUEST',
                    message: error.detail || 'Failed to create member'
                })
            }
            const data = await response.json()
            logger.info('Successfully created member', { id: data.id })
            return data
        }),

    update: protectedProcedure
        .input(
            z.object({
                id: z.number(),
                name: z.string().optional(),
                email: z.string().email().optional(),
                address: z.string().optional()
            })
        )
        .mutation(async ({ input }): Promise<Member> => {
            logger.info('Updating member', { input })
            const { id, ...data } = input
            const response = await fetch(`${API_URL}/api/members/${id}`, {
                method: 'PUT',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify(data)
            })
            if (!response.ok) {
                const error = await response.json()
                logger.error('Failed to update member', { input, error })
                throw new TRPCError({
                    code: 'BAD_REQUEST',
                    message: error.detail || 'Failed to update member'
                })
            }
            const updatedMember = await response.json()
            logger.info('Successfully updated member', { id })
            return updatedMember
        }),

    delete: protectedProcedure.input(z.number()).mutation(async ({ input }): Promise<Member> => {
        logger.info('Deleting member', { id: input })
        const response = await fetch(`${API_URL}/api/members/${input}`, {
            method: 'DELETE'
        })
        if (!response.ok) {
            const error = await response.json()
            logger.error('Failed to delete member', { id: input, error })
            throw new TRPCError({
                code: 'BAD_REQUEST',
                message: error.detail || 'Failed to delete member'
            })
        }
        const deletedMember = await response.json()
        logger.info('Successfully deleted member', { id: input })
        return deletedMember
    }),

    payDebt: protectedProcedure
        .input(
            z.object({
                id: z.number(),
                amount: z.number().positive()
            })
        )
        .mutation(async ({ input }): Promise<Member> => {
            const response = await fetch(`${API_URL}/api/members/${input.id}/pay-debt?amount=${input.amount}`, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' }
            })
            if (!response.ok) {
                const error = await response.json()
                throw new TRPCError({
                    code: 'BAD_REQUEST',
                    message: error.detail || 'Failed to process payment'
                })
            }
            return response.json()
        }),

    clearDebt: protectedProcedure.input(z.number()).mutation(async ({ input }) => {
        logger.info('Clearing debt for member', { id: input })
        const response = await fetch(`${API_URL}/api/members/${input}/clear-debt`, {
            method: 'POST'
        })
        if (!response.ok) {
            const error = await response.json()
            logger.error('Failed to clear debt', { id: input, error })
            throw new TRPCError({
                code: 'BAD_REQUEST',
                message: error.detail || 'Failed to clear debt'
            })
        }
        const result = await response.json()
        logger.info('Successfully cleared debt for member', { id: input })
        return result
    })
})

File: server/routers/member.py
from fastapi import APIRouter, HTTPException, Query
from typing import Optional
from services.member_service import MemberService
from exceptions import LibraryException, ResourceNotFoundException, InvalidOperationException
from models import Member, PaginatedResponse, MemberCreate

router = APIRouter(prefix="/api/members", tags=["members"])
member_service = MemberService()

@router.get("")
async def get_members(
    search: Optional[str] = None,
    page: int = Query(default=1, ge=1),
    limit: int = Query(default=20, ge=1)
):
    try:
        return await member_service.get_members(search, page, limit)
    except LibraryException as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail="Internal server error")

@router.get("/{member_id}")
async def get_member_by_id(member_id: int):
    try:
        return await member_service.get_member_by_id(member_id)
    except ResourceNotFoundException as e:
        raise HTTPException(status_code=404, detail=str(e))
    except LibraryException as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail="Internal server error")

@router.post("")
async def create_member(member: MemberCreate):
    try:
        return await member_service.create_member(member)
    except LibraryException as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail="Internal server error")

@router.put("/{member_id}")
async def update_member(member_id: int, member: MemberCreate):
    try:
        return await member_service.update_member(member_id, member)
    except ResourceNotFoundException as e:
        raise HTTPException(status_code=404, detail=str(e))
    except LibraryException as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail="Internal server error")

@router.delete("/{member_id}")
async def delete_member(member_id: int):
    try:
        return await member_service.delete_member(member_id)
    except ResourceNotFoundException as e:
        raise HTTPException(status_code=404, detail=str(e))
    except InvalidOperationException as e:
        raise HTTPException(status_code=400, detail=str(e))
    except LibraryException as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail="Internal server error")

@router.post("/{member_id}/pay-debt")
async def pay_debt(member_id: int, amount: float = Query(..., gt=0)):
    try:
        return await member_service.pay_debt(member_id, amount)
    except ResourceNotFoundException as e:
        raise HTTPException(status_code=404, detail=str(e))
    except InvalidOperationException as e:
        raise HTTPException(status_code=400, detail=str(e))
    except LibraryException as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail="Internal server error")

@router.post("/{member_id}/clear-debt")
async def clear_debt(member_id: int):
    try:
        return await member_service.clear_debt(member_id)
    except ResourceNotFoundException as e:
        raise HTTPException(status_code=404, detail=str(e))
    except LibraryException as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail="Internal server error") 